hashmap/src

```diff
@@ -0,0 +1,248 @@
+use std::collections::HashMap;
+
+//mod tej_serializer;
+pub trait MeasurementCollector {
+    type Error;
+    type Data;
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error>;
+    fn measurement(
+        &mut self,
+        group: Option<&str>,
+        name: &str,
+        value: f64,
+    ) -> Result<(), Self::Error>;
+}
+#[derive(Debug)]
+pub struct ThinEdgeJsonMap {
+    pub timestamp: String,
+    pub values: HashMap<String, Measurement>,
+}
+#[derive(Debug)]
+pub enum Measurement {
+    Single(f64),
+    Multi(HashMap<String, f64>),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ThinEdgeJsonMapError {
+    #[error("Duplicated measurement: {0}")]
+    DuplicatedMeasurement(String),
+
+    #[error("Duplicated measurement: {0}.{1}")]
+    DuplicatedSubMeasurement(String, String),
+}
+
+impl ThinEdgeJsonMap {
+    fn new() -> Self {
+        Self {
+            timestamp: "4-20-2021".into(),
+            values: HashMap::new(),
+        }
+    }
+}
+
+impl MeasurementCollector for ThinEdgeJsonMap {
+    type Error = ThinEdgeJsonMapError;
+    type Data = ThinEdgeJsonMap;
+
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error> {
+        self.timestamp = value;
+        Ok(())
+    }
+
+    fn measurement(
+        &mut self,
+        group: Option<&str>,
+        name: &str,
+        value: f64,
+    ) -> Result<(), Self::Error> {
+        let key = name.to_owned();
+
+        match group {
+            None => {
+                self.values.insert(key, Measurement::Single(value));
+                return Ok(());
+            }
+            Some(group) => {
+                let key = group.to_owned();
+
+                if !self.values.contains_key(&key) {
+                    self.values
+                        .insert(key.clone(), Measurement::Multi(HashMap::new()));
+                }
+
+                let group = match self.values.get_mut(&key) {
+                    Some(Measurement::Multi(group)) => group,
+                    _ => {
+                        return Err(ThinEdgeJsonMapError::DuplicatedMeasurement(key));
+                    }
+                };
+
+                let sub_key = name.to_owned();
+                group.insert(sub_key, value);
+                Ok(())
+            }
+        }
+    }
+}
+
+use std::io::Write;
+pub trait MeasurementConsumer {
+    type Error;
+    type Data;
+    fn start(&mut self) -> Result<(), Self::Error>;
+    fn end(&mut self) -> Result<(), Self::Error>;
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error>;
+    fn measurement(&mut self, name: &str, value: f64) -> Result<(), Self::Error>;
+    fn start_group(&mut self, group: &str) -> Result<(), Self::Error>;
+    fn end_group(&mut self) -> Result<(), Self::Error>;
+}
```

Comment:
The names don't highlight the different between the two traits introduced here (so called `MeasurementConsumer` and `MeasurementCollector`) .

* Both are consuming measurements pushed by some external produced (the producer knows what are the measurements, the consumer knows what to do with them).
* The second trait consumes the measurements group by group, While the first can consume measurements in any order independently of their group.
* I would prefer to name the first one a `MeasurementConsumer`, the second one a `GroupedMeasurementConsumer`

Reply:
I actually merged both the traits into one in the recent commit. Now only MeaurementConsumer trait exists. Though some functions are not needed in producer cases. Needed some dummy implementation. Do you prefer two different traits?